service/history/queue/cross_cluster_queue_processor_base.go

```diff
@@ -0,0 +1,688 @@
+			}
+		},
+
+		backoffTimer:  make(map[int]*time.Timer),
+		shouldProcess: make(map[int]bool),
+
+		notifyCh:  make(chan struct{}, 1),
+		processCh: make(chan struct{}, 1),
+	}
+
+	return crossClusterQueueProcessorBase
+}
+
+func (c *crossClusterQueueProcessorBase) Start() {
+	if !atomic.CompareAndSwapInt32(&c.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	c.logger.Info("Cross cluster queue processor state changed", tag.LifeCycleStarting)
+	defer c.logger.Info("Cross cluster queue processor state changed", tag.LifeCycleStarted)
+
+	c.redispatcher.Start()
+
+	// trigger an initial load of tasks
+	c.notifyAllQueueCollections()
+
+	c.shutdownWG.Add(1)
+	go c.processLoop()
+}
+
+func (c *crossClusterQueueProcessorBase) Stop() {
+	if !atomic.CompareAndSwapInt32(&c.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	c.logger.Info("Cross cluster queue processor state changed", tag.LifeCycleStopping)
+	defer c.logger.Info("Cross cluster queue processor state changed", tag.LifeCycleStopped)
+
+	close(c.shutdownCh)
+	c.processingLock.Lock()
+	for _, timer := range c.backoffTimer {
+		timer.Stop()
+	}
+	for level := range c.shouldProcess {
+		c.shouldProcess[level] = false
+	}
+	c.processingLock.Unlock()
+
+	if success := common.AwaitWaitGroup(&c.shutdownWG, time.Minute); !success {
+		c.logger.Warn("", tag.LifeCycleStopTimedout)
+	}
+
+	c.redispatcher.Stop()
+}
+
+func (c *crossClusterQueueProcessorBase) processLoop() {
+	defer c.shutdownWG.Done()
+
+	updateAckTimer := time.NewTimer(backoff.JitDuration(
+		c.options.UpdateAckInterval(),
+		c.options.UpdateAckIntervalJitterCoefficient(),
+	))
+	defer updateAckTimer.Stop()
+
+	maxPollTimer := time.NewTimer(backoff.JitDuration(
+		c.options.MaxPollInterval(),
+		c.options.MaxPollIntervalJitterCoefficient(),
+	))
+	defer maxPollTimer.Stop()
+
+	var ackLevel int64
+	lastPollTime := time.Now()
+processorPumpLoop:
+	for {
+		select {
+		case <-c.shutdownCh:
+			break processorPumpLoop
+		case <-c.notifyCh:
+			c.notifyAllQueueCollections()
+		case <-updateAckTimer.C:
+			processFinished, err := c.updateAckLevel()
+			if err == shard.ErrShardClosed || (err == nil && processFinished) {
+				go c.Stop()
+				break processorPumpLoop
+			}
+			ackLevel, err = c.completeAckTasks(ackLevel)
+			if err != nil {
+				c.logger.Error("failed to complete ack tasks", tag.Error(err))
+				c.metricsScope.IncCounter(metrics.TaskBatchCompleteFailure)
+			}
+			updateAckTimer.Reset(backoff.JitDuration(
+				c.options.UpdateAckInterval(),
+				c.options.UpdateAckIntervalJitterCoefficient(),
+			))
+		case <-maxPollTimer.C:
+			c.notifyAllQueueCollections()
+			if time.Now().After(lastPollTime.Add(c.options.ValidationInterval())) {
+				// No polling is happening and loop all outstanding task and validate task status
+				c.validateOutstandingTasks()
+				lastPollTime = time.Now()
+			}
```

Comment:
I am thinking if we can do this validation on task load and when `FailoverDomain()` method is called on the Processor. Then I guess we don't need to perform this check periodically and can react more quickly upon domain failover?